src/utilities/include/vec.h

```diff
@@ -0,0 +1,318 @@
+// Copyright 2021 The Manifold Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <exception>
+#if TRACY_ENABLE && TRACY_MEMORY_USAGE
+#include "tracy/Tracy.hpp"
+#else
+#define TracyAllocS(ptr, size, n) (void)0
+#define TracyFreeS(ptr, n) (void)0
+#endif
+
+// #include "optional_assert.h"
+#include "par.h"
+#include "public.h"
+
+namespace manifold {
+
+/** @addtogroup Private
+ *  @{
+ */
+template <typename T>
+class Vec;
+
+/**
+ * View for Vec, can perform offset operation.
+ * This will be invalidated when the original vector is dropped or changes
+ * length.
+ */
+template <typename T>
+class VecView {
+ public:
+  using Iter = T *;
+  using IterC = const T *;
+
+  VecView(const VecView &other) {
+    ptr_ = other.ptr_;
+    size_ = other.size_;
+  }
+
+  VecView &operator=(const VecView &other) {
+    ptr_ = other.ptr_;
+    size_ = other.size_;
+    return *this;
+  }
+
+  // allows conversion to a const VecView
+  operator VecView<const T>() const { return {ptr_, size_}; }
+
+  inline const T &operator[](int i) const {
+    if (i < 0 || i >= size_) throw std::out_of_range("Vec out of range");
```

Comment:
Does this really not make a perf difference? I'm surprised, but that would be great.

Reply:
yeah, there is a very tiny performance difference (50ms/3000ms)

this kind of checks are fast because the compiler can often optimize them away when inlining, e.g. in a loop and the compiler can prove that the bound must hold
even when the compiler cannot optimize away this, it is generally in the cold path and does not contribute much to the latency.

in fact, rust does bound check for safe code, and rust can write fast programs without using unsafe.